pkg/envoy/xds_server.go

```diff
@@ -1398,58 +1421,55 @@ func getWildcardNetworkPolicyRule(selectors policy.L7DataMap) *cilium.PortNetwor
 				// No remote policies would match this rule. Discard it.
 				return nil
 			}
-			// convert from []uint32 to []uint64
-			remotePolicies := make([]uint32, len(selections))
-			for i, id := range selections {
-				remotePolicies[i] = uint32(id)
-			}
 			return &cilium.PortNetworkPolicyRule{
-				RemotePolicies: remotePolicies,
+				RemotePolicies: selections.AsUint32Slice(),
 			}
 		}
 	}
 
-	// Use map to remove duplicates
-	remoteMap := make(map[uint32]struct{})
+	// Get selections for each selector and count how many there are
+	sels := make([][]uint32, 0, len(selectors))
 	wildcardFound := false
+	var count int
 	for sel, l7 := range selectors {
 		if sel.IsWildcard() {
 			wildcardFound = true
 			break
 		}
 
-		for _, id := range sel.GetSelections() {
-			remoteMap[uint32(id)] = struct{}{}
-		}
-
 		if l7.IsRedirect() {
 			// Issue a warning if this port-0 rule is a redirect.
 			// Deny rules don't support L7 therefore for the deny case
 			// l7.IsRedirect() will always return false.
 			log.Warningf("L3-only rule for selector %v surprisingly requires proxy redirection (%v)!", sel, *l7)
 		}
+
+		selections := sel.GetSelections()
+		if len(selections) == 0 {
+			continue
+		}
+		count += len(selections)
+		sels = append(sels, selections.AsUint32Slice())
 	}
 
+	var remotePolicies []uint32
+
 	if wildcardFound {
 		// Optimize the policy if the endpoint selector is a wildcard by
 		// keeping remote policies list empty to match all remote policies.
-		remoteMap = nil
-	} else if len(remoteMap) == 0 {
+	} else if count == 0 {
 		// No remote policies would match this rule. Discard it.
 		return nil
+	} else {
+		// allocate slice and copy selected identities
+		remotePolicies = make([]uint32, count)
+		var i int
+		for _, selections := range sels {
+			i += copy(remotePolicies[i:], selections)
+		}
```

Comment:
```suggestion
		remotePolicies = make([]uint32, 0, count)
		for _, selections := range sels {
			remotePolicies = append(remotePolicies, selections...)
		}
```

https://go.dev/doc/effective_go#append

Reply:
Thanks for this! I had assumed copy() be faster, but it turns out it is not! Changed as suggested.